File: Kakao/2018_프렌즈4블록.swift
// Time Complexity: O(m^3*n)

var numOfPops: Int = 0

// O(mn)
func checkBlocksToBurst(gameBoard: [[String]], blocksToBurst: inout [[Bool]], bursted: inout Bool) {
    bursted = false
    for i in 0..<gameBoard.count-1 {
        for j in 0..<gameBoard[0].count-1 {
            if gameBoard[i][j] != "POP" && gameBoard[i][j] == gameBoard[i][j+1] && gameBoard[i][j] == gameBoard[i+1][j] && gameBoard[i][j] == gameBoard[i+1][j+1] {
                blocksToBurst[i][j] = true
                blocksToBurst[i][j+1] = true
                blocksToBurst[i+1][j] = true
                blocksToBurst[i+1][j+1] = true
                bursted = true
            }
        }
    }
}

//O(mn)
func burstBlocks(gameBoard: inout [[String]], blocksToBurst: [[Bool]]) {
    for i in 0..<gameBoard.count {
        for j in 0..<gameBoard[0].count {
            if blocksToBurst[i][j] {
                gameBoard[i][j] = "POP"
                numOfPops += 1
            }
        }
    }
}

O(m^2*n)
func dropBlocks(gameBoard: inout [[String]])
{
    for col in 0..<gameBoard[0].count {
        for _ in 0..<gameBoard.count-1 {
            for row in (1..<gameBoard.count) {
                if gameBoard[row][col] == "POP" && gameBoard[row-1][col] != "POP" {
                    gameBoard[row][col] = gameBoard[row-1][col]
                    gameBoard[row-1][col] = "POP"
                }
            }
        }
    }
}

func solution(_ m:Int, _ n:Int, _ board:[String]) -> Int {
    var gameBoard = board.map({$0.map({String($0)})})
    var bursted = true
    // repeat 최대 m/2회 반복.
    repeat {
        var blocksToBurst = [[Bool]].init(repeating: [Bool].init(repeating: false, count: n), count: m)
        checkBlocksToBurst(gameBoard: gameBoard, blocksToBurst: &blocksToBurst, bursted: &bursted)
        burstBlocks(gameBoard: &gameBoard, blocksToBurst: blocksToBurst)
        dropBlocks(gameBoard: &gameBoard)
    } while bursted
    
    return numOfPops
}

File: Kakao/2018_프렌즈4블록.cpp


#include <string>
#include <vector>
#include <iostream>

using namespace std;

int Check(int m, int n, vector<string> &board)
{
    int cnt=0;
    vector<vector<bool>> check(m,vector<bool>(n,false));
    
    for(int i=0;i<m-1;i++)
    {
        for(int j=0;j<n-1;j++)
        {
            if(isalpha(board[i][j]) && board[i][j]==board[i+1][j] && board[i][j]==board[i][j+1] && board[i][j]==board[i+1][j+1])
            {
                check[i][j] = check[i][j+1] = check[i+1][j] = check[i+1][j+1] = true;
            }
        }
    }
    
    for(int i=0;i<m;i++)
    {
        for(int j=0;j<n;j++)
        {
            if(check[i][j])  
            {
                board[i][j] = '!';
                cnt++;
            }
        }
    }
    return cnt;
}

void trickleDown(int m, int n, vector<string> &board, int col)
{
    for(int i=m-1;i>=0;i--)
    {
        if(board[i][col]=='!' && i>0)
        {
            for(int j=i-1;j>=0;j--)
            {
                if(isalpha(board[j][col]))  
                {
                    swap(board[i][col],board[j][col]);
                    break;
                }
            }
        }
    }
}

int Pop(int m, int n, vector<string> &board)
{
    int cnt = Check(m,n,board);
    for(int j=0;j<n;j++)
    {
        trickleDown(m,n,board,j);
    }
    return cnt;
}

int solution(int m, int n, vector<string> board) {
    int answer = 0;
    int cnt=0;
    
    while(cnt=Pop(m,n,board))
    {
        answer += cnt;
    }
    
    return answer;
}
